lib/service_actor.rb

```diff
@@ -68,6 +89,20 @@ def after; end
 
   # :nodoc:
   def run
+    # Oh, I see, so everyone gonna plug itself on those methods and call super.
+    # That is tricky to know which module gonna go first. Usually to achieve
+    # that behavior, I'm using an instance variable at the class level holding
+    # a list of before_hooks and after_hooks (`Proc`). This way for each class
+    # that ordered list is easily available for inspection.
+    #
+    # Did you consider that approach?
+    #
+    # The good part is that the included/prepended modules don't have to
+    # redefine methods and call super, but 'simply' define the DSL keywords and
+    # interact with the base API that register before_hooks and after_hooks.
+    #
+    # Another good part is that the same codepath is always executed, only the
+    # data changes, the content of those before_hooks and after_hooks.
```

Comment:
I have considered that approach but haven't tried it here to see how it would change the code.

I tend to favor overriding and calling `super`, since I find it rather natural in Ruby and not have to define a DSL. To me using instance methods is simpler than procs.

The list of modules is also easily available for inspection, since it is in the class  hierarchy. But I could be tempted to try it out.

Not sure about the fact that "the same codepath is always executed", though. We could debate that's the same for both solutions :)